### PR TITLE
Use nodes list in JSON

### PR DESCRIPTION
**Pull request by bergie**
Should help with auto-saving when running a partial network (some participants turned off)

### CODE REVIEW

**Comment by jonnor** [comment]
If we also add all the discovered roles from `@participants`, how is this different from reading it from there, like before?

File: src/coordinator.coffee [hunk 1]
```diff
@@ -378,19 +386,19 @@ class Coordinator extends EventEmitter
...
 
     if action == 'added'
-      id = participant.id
+      role = participant.role
       # inbound
       for port in participant.inports
         matches = findConnectedPorts 'outports', port
         for m in matches
-          e = [m.part.id, m.port.id, id, port.id]
+          e = [m.part.role, m.port.id, role, port.id]
```
Comment (jonnor) [inline src/coordinator.coffee]: Good catch. This could actually have caused multiple connections

**Comment by jonnor** [comment]
Replying to myself, it helps in the case where we add a node from the initial graph. If only using `@participants` the discovery comes in much later (or not at all in case something is temporarily down), so they can get lost when serializing again. `@nodes` are pre-added and thus does not have that problem

**Comment by bergie** [comment]
Yes, and since we also add discovered additional participants to `@nodes`, the nodes list becomes a combination of:

* Things defined in the initial graph (whether running or not)
* Any additional participants that were announcing themselves (newly-added devices etc)

**Pull request merged**